c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1474,413 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+};
+
+}  // namespace KJ_COROUTINE_STD_NAMESPACE
+
+// Now when the compiler sees our `connectToService()` coroutine above, it default-constructs a
+// `coroutine_traits<Promise<Own<AsyncIoStream>>, Network&>::promise_type`, or
+// `kj::_::Coroutine<Own<AsyncIoStream>>`.
+//
+// The implementation object lives in the heap-allocated coroutine frame. It gets destroyed and
+// deallocated when the frame does.
+
+namespace kj::_ {
+
+namespace stdcoro = KJ_COROUTINE_STD_NAMESPACE;
+
+template <typename Self, typename T>
+class CoroutineBase;
+// CRTP mixin, covered later.
+
+template <typename T>
+class Coroutine: public CoroutineBase<Coroutine<T>, T>,
+                 private PromiseNode,
+                 private Event,
+                 private Disposer {
+  // The standard calls this the `promise_type` object. We can call this the "coroutine
+  // implementation object" since the word promise means different things in KJ and std styles. This
+  // is where we implement how a `kj::Promise<T>` is returned from a coroutine, and how that promise
+  // is later fulfilled. We also fill in a few lifetime-related details.
+  //
+  // The implementation object is also where we can customize memory allocation of coroutine frames,
+  // by implementing a member `operator new(size_t, Args...)` (same `Args...` as in
+  // coroutine_traits).
+  //
+  // We can also customize how await-expressions are transformed within `kj::Promise<T>`-based
+  // coroutines by implementing an `await_transform(P)` member function, where `P` is some type for
+  // which we want to implement co_await support, e.g. `kj::Promise<U>`. This feature is useful
+  // to provide a more optimized `kj::EventLoop` integration when the coroutine's return type and
+  // the await-expression's type are both `kj::Promise` instantiations.
+
+public:
+  using Handle = stdcoro::coroutine_handle<Coroutine<T>>;
+
+  Coroutine() = default;
+  ~Coroutine() noexcept(false) { readMaybe(maybeDisposalResults)->destructorRan = true; }
+  KJ_DISALLOW_COPY(Coroutine);
+
+  Promise<T> get_return_object() {
+    // Called after coroutine frame construction and before initial_suspend() to create the
+    // coroutine's return object. `this` itself lives inside the coroutine frame, and we arrange for
+    // the returned Promise<T> to own `this` via a custom Disposer and by always leaving the
+    // coroutine in a suspended state.
+    return PromiseNode::to<Promise<T>>(Own<PromiseNode>(this, *this));
+  }
+
+private:
+  // -------------------------------------------------------
+  // Disposer implementation
+
+  void disposeImpl(void* pointer) const override {
+    KJ_IASSERT(pointer == this);
+
+    // const_cast okay -- every Own<PromiseNode> that we build in get_return_object() uses itself
+    // as the disposer, thus every disposer is unique and there are no thread-safety concerns.
+    const_cast<Coroutine&>(*this).destroy();
+  }
+
+  void destroy() {
```

Comment:
There's a lot of logic here that doesn't seem like it needs to be specialized for every promise type. Should we factor out a non-template base class to put these in? Also for `Awaiter`.

Reply:
https://github.com/capnproto/capnproto/pull/1104/commits/f97587b67a1423e4d188d1b6adb7658d8c906aa5, https://github.com/capnproto/capnproto/pull/1104/commits/d37e57cfa6a3b3ccddffe89d0a483f2b32c65904, and https://github.com/capnproto/capnproto/pull/1104/commits/d8fd23da384819150dcb6337e6e2d9a1adf865db